File: editor/src/vocabulary.js
// The suggestion lists behind every dropdown in the material editor, and the
// set of constitutive kinds the validator recognises.
//
// `assets/vocabulary.json` is read by gm_core as well, so the editor cannot
// offer a constitutive kind that `gm validate` would then warn about. It is a
// suggestion list, not a schema: the format keeps property names, units, soil
// classes and kinds open, and anything typed outside these lists round-trips
// untouched.
import vocabulary from "../../assets/vocabulary.json";

export const soilClasses = vocabulary.soilClasses;
export const materialProperties = vocabulary.properties;
export const constitutiveKinds = vocabulary.constitutiveKinds;

export const KNOWN_CONSTITUTIVE_KINDS = new Set(constitutiveKinds.map((kind) => kind.kind));

// Every unit anyone suggests anywhere, for a property this file has never heard
// of: better a long list than none.
export const allUnits = [...new Set([
  ...materialProperties.flatMap((property) => property.units),
  ...constitutiveKinds.flatMap((kind) => kind.parameters.map((parameter) => parameter.units).flat()),
])];

const propertyIndex = new Map(materialProperties.map((property) => [property.key, property]));
const kindIndex = new Map(constitutiveKinds.map((kind) => [kind.kind, kind]));

export function constitutiveKind(kind) {
  return kindIndex.get(kind);
}

export function kindParameters(kind) {
  return kindIndex.get(kind)?.parameters ?? [];
}

// The terms suggested for a material's `properties` map, or for one
// constitutive model's `parameters` map when `kind` is given.
export function terms(kind) {
  return kind === undefined ? materialProperties : kindParameters(kind);
}

export function term(key, kind) {
  return kind === undefined ? propertyIndex.get(key) : kindParameters(kind).find((p) => p.key === key);
}

// Units to offer for a term. An unrecognised term gets the full list rather
// than an empty one, because a custom property still needs a unit.
export function unitsFor(key, kind) {
  const units = term(key, kind)?.units;
  return units?.length ? units : allUnits;
}

// The unit a term is normally given in — used when a property is first chosen,
// so the common case needs no typing and cannot be mistyped.
export function defaultUnit(key, kind) {
  return term(key, kind)?.units[0];
}

export function label(key, kind) {
  return term(key, kind)?.label;
}

// The first suggested term not already present, for "+ Add". Falls back to a
// placeholder key when a material already carries all of them.
export function nextTerm(existingKeys, kind) {
  const used = new Set(existingKeys);
  return terms(kind).find((candidate) => !used.has(candidate.key));
}

File: crates/gm-core/src/vocabulary.rs
//! The advisory vocabulary shared with the browser editor.
//!
//! `assets/vocabulary.json` is what the editor offers in its dropdowns:
//! property names, unit strings, soil classes and constitutive kinds. It is a
//! suggestion list, not a schema — the format keeps all four open on purpose
//! (see `docs/format.md`) and anything outside it round-trips untouched.
//!
//! One part of it *is* load-bearing here: the set of constitutive kinds. If the
//! editor offered a kind that [`crate::validate`] then warned about, the file
//! would look wrong the moment it reached `gm commit`. Both sides read the same
//! list rather than keeping a copy each.

use serde::Deserialize;
use std::collections::BTreeSet;
use std::sync::LazyLock;

const SOURCE: &str = include_str!("../../../assets/vocabulary.json");

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct Vocabulary {
    soil_classes: Vec<String>,
    properties: Vec<Term>,
    constitutive_kinds: Vec<Kind>,
}

#[derive(Debug, Deserialize)]
pub struct Term {
    pub key: String,
    pub label: String,
    pub units: Vec<String>,
}

#[derive(Debug, Deserialize)]
pub struct Kind {
    pub kind: String,
    pub label: String,
    #[serde(default)]
    pub parameters: Vec<Term>,
}

static VOCABULARY: LazyLock<Vocabulary> = LazyLock::new(|| {
    serde_json::from_str(SOURCE).expect("assets/vocabulary.json is built in and must parse")
});

static KINDS: LazyLock<BTreeSet<&'static str>> = LazyLock::new(|| {
    VOCABULARY
        .constitutive_kinds
        .iter()
        .map(|k| k.kind.as_str())
        .collect()
});

/// Constitutive kinds this build understands well enough to check parameters
/// for. Anything else is carried through untouched and warned about once.
pub fn knows_constitutive_kind(kind: &str) -> bool {
    KINDS.contains(kind)
}

pub fn constitutive_kinds() -> &'static [Kind] {
    &VOCABULARY.constitutive_kinds
}

pub fn properties() -> &'static [Term] {
    &VOCABULARY.properties
}

pub fn soil_classes() -> &'static [String] {
    &VOCABULARY.soil_classes
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_vocabulary_parses_and_is_not_empty() {
        assert!(!properties().is_empty());
        assert!(!soil_classes().is_empty());
        assert!(!constitutive_kinds().is_empty());
    }

    /// The kinds `validate` writes parameter-specific checks for must be in the
    /// list, or those checks would never run and the editor would not offer the
    /// kind that triggers them.
    #[test]
    fn kinds_with_parameter_checks_are_offered() {
        for kind in ["mohr-coulomb", "undrained-tresca"] {
            assert!(
                knows_constitutive_kind(kind),
                "{kind} is not in the vocabulary"
            );
        }
        assert!(!knows_constitutive_kind("something-newer"));
    }

    /// Every suggested term has to name a unit, because a parameter without one
    /// is a warning the moment it is saved.
    #[test]
    fn every_suggested_term_has_a_unit() {
        let terms = properties().iter().chain(
            constitutive_kinds()
                .iter()
                .flat_map(|k| k.parameters.iter()),
        );
        for term in terms {
            assert!(!term.units.is_empty(), "{} suggests no unit", term.key);
            assert!(!term.label.is_empty(), "{} has no label", term.key);
        }
    }
}
